### PR TITLE
Set minimum required standard to C++17

### LINKED ISSUE #26464: C++17 migration umbrella issue
It was agreed that we would aim to move to C++17 after release 4.1 was completed. This is an umbrella issue to track the requirements.

Tasks:

- [x] Move to VS2019 on Windows (#26217)
  - [x] 2 ISIS builders also need to move to Windows 10 (#26219)
  - [x] Changes required to build with VS2019 merged to `master` but still build with 2017
  - [x] Update build scripts to build with 2019
- [x] Drop support for Ubuntu 16.04
  - [x] Builders need to move
- [x] Move to macOS High Sierra as the minimum supported system (#26261) 
  - [x] Builders need to move
- [ ] Update pull request config (if necessary)
- [x] Update clang-format

### PR DESCRIPTION
**Pull request by martyngigg**
**Description of work.**

Require a compiler with support for the C++17 language standard as a minimum. Note that this does not require a C++17 compliant std library implementation. For example macOS High Sierra has Apple Clang 10 (based on LLVM clang 6) that is C++17 language complete but libcxx does not have std::filesystem yet.

Despite this, access to the C++17 language features is still a win. 

**To test:**

Builds should pass.

Fixes #26464

*This does not require release notes* because **it is maintenance.**

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by DavidFair** [comment]
It looks like we have a few deprecated call to `std::mem_fun_ref` which can be directly replaced with `std::mem_fn_ref` as it's encapsulated to the class implementation

Is this something which would be tackled in this PR or in a separate one which will block this?

**Comment by martyngigg** [comment]
Thanks, David. I've fixed the minor `mem_fun` issues and rebased this on to of my PR to move boost forward as apparently we need that too. Would you have time to review #26480?

**Comment by DavidFair** [comment]
Okay, I'll review that today first - thanks for the heads up

**Comment by martyngigg** [comment]
It looks like clang is far better at actually removing stuff that is not supposed to be there. I'll compile on mac to figure the rest out.

**Comment by martyngigg** [comment]
> Okay, I'll review that today first - thanks for the heads up

Thanks!

**Comment by martyngigg** [comment]
And it looks like we need to upgrade tbb on Windows too :disappointed:

File: Framework/API/src/FileFinder.cpp [hunk 1]
```diff
@@ -207,8 +205,8 @@ FileFinderImpl::toInstrumentAndNumber(const std::string &hint) const {
     runPart = hint;
   } else {
     /// Find the last non-digit as the instrument name can contain numbers
-    std::string::const_reverse_iterator it = std::find_if(
-        hint.rbegin(), hint.rend(), std::not1(std::ptr_fun(isdigit)));
+    std::string::const_reverse_iterator it =
+        std::find_if(hint.rbegin(), hint.rend(), std::not_fn(isdigit));
```
Comment (DavidFair) [inline Framework/API/src/FileFinder.cpp]: Could we take this opportunity to replace all of the `std::not_fn` with lambdas. That way they don't have to lookup `std::not_fn` if they aren't familiar:

`std::find_if(hint.rbegin(), hint.rend(), [](const char c){return !isdigit(c);});`

File: Framework/Beamline/test/ComponentInfoTest.h [hunk 2]
```diff
@@ -755,7 +755,7 @@ class ComponentInfoTest : public CxxTest::TestSuite {
     TS_ASSERT(compInfo.relativePosition(subComponentIndex)
                   .isApprox(subCompPosition - rootPosition));
 
-    const auto diffPos =
+    const Vector3d diffPos =
```
Comment (DavidFair) [inline Framework/Beamline/test/ComponentInfoTest.h]: Could we add a comment explaining why `auto` is not a good idea, to save this bug creeping back in?

File: Framework/DataObjects/test/Histogram1DTest.h [hunk 3]
```diff
@@ -170,16 +171,22 @@ class Histogram1DTest : public CxxTest::TestSuite {
   }
   void testrangeexceptionX() {
     h.setPoints(x1);
+    MSVC_DIAG_OFF(4834)
```
Comment (DavidFair) [inline Framework/DataObjects/test/Histogram1DTest.h]: Can we comment about the `nodiscard`

File: Framework/Geometry/src/Math/Acomp.cpp [hunk 4]
```diff
@@ -214,8 +214,8 @@ Complementary subtraction is by making A-B == A*B'
 
   // Have DNF parts and will return the same...
   // Sort the components of the list
-  std::for_each(Fparts.begin(), Fparts.end(), std::mem_fun_ref(&Acomp::Sort));
-  std::for_each(Gparts.begin(), Gparts.end(), std::mem_fun_ref(&Acomp::Sort));
+  std::for_each(Fparts.begin(), Fparts.end(), std::mem_fn(&Acomp::Sort));
```
Comment (DavidFair) [inline Framework/Geometry/src/Math/Acomp.cpp]: Same with `sort` and a lambda

**Review by DavidFair — CHANGES_REQUESTED** [review]

File: Framework/Beamline/test/ComponentInfoTest.h (hunk 2 above)
Comment (martyngigg) [reply Framework/Beamline/test/ComponentInfoTest.h]: :+1:

File: Framework/DataObjects/test/Histogram1DTest.h (hunk 3 above)
Comment (martyngigg) [reply Framework/DataObjects/test/Histogram1DTest.h]: :+1:

File: Framework/API/src/FileFinder.cpp (hunk 1 above)
Comment (martyngigg) [reply Framework/API/src/FileFinder.cpp]: Do you not think `not_fn(isdigit)` is more readable? I have to read further with the lambda and make sure I don't miss the `!` in there.

File: Framework/Geometry/src/Math/Acomp.cpp (hunk 4 above)
Comment (martyngigg) [reply Framework/Geometry/src/Math/Acomp.cpp]: So this is not actually sorting `Fparts`/`Gparts` (that is done below). It's calling `Acomp::Sort` on each entry in `Fparts`/`Gparts`. I'm not sure a lambda clarifies this much?

File: Framework/API/src/FileFinder.cpp (hunk 1 above)
Comment (DavidFair) [reply Framework/API/src/FileFinder.cpp]: 🤔 Good point - it's finding a balance between dev's infrequently seeing `std::not_fn` and the boilerplate of a lambda

I think I'll agree with you on this one, as once you've looked up the various memory references the STL gives you it's obvious to read

**Review by DavidFair — APPROVED** [review]
I'm happy with this, lets :shipit: once the tests re-run

**Review by DanNixon — APPROVED** [review]
:+1:

**Comment by DanNixon** [comment]
It wasn't worth not merging for but the node about Eigen should be somewhere more generic, that rule (no `auto`) applies top all uses of Eigen.

**Comment by martyngigg** [comment]
Yes. I just don't know where. Any suggestions?

**Comment by DanNixon** [comment]
No idea. Maybe just a PSA in Slack. The Eigen docs regarding the issue are easy to find by searching "eigen auto".

**Comment by DavidFair** [comment]
If it really became a problem, we could create an adapter class for Eigen which guarantees that even if a dev uses `auto` the adapter will manually coerce the type 

However, I think that just making the two or three people who actively touch the Eigen code aware would be a better investment of time

**Pull request merged**